src/psc_datalogger/connection.py

```diff
@@ -0,0 +1,410 @@
+
+            results = self.query_instruments()
+
+            if any(x == self.ERROR_STRING for x in results):
+                simple_timestamp = results[0].isoformat(sep=" ", timespec="seconds")
+                logging.error("Unable to read data")
+                self.error.emit(f"Unable to read data {simple_timestamp}")
+            else:
+                self.query_complete.emit(results[0])
+
+            logging.info(
+                f"Data read: {str(results[0])} {results[1]} {results[2]} {results[3]}"
+            )
+
+            self.writer.write(
+                timestamp=results[0],
+                ins_1=results[1],
+                ins_2=results[2],
+                ins_3=results[3],
+            )
+
+    def query_instruments(self) -> Tuple[datetime, str, str, str]:
+        """Query the instruments and return the timestamp followed by three instrument
+        readings."""
+
+        with self.lock:
+            measurement_time = datetime.now()
+
+            measurements: List[str] = []
+            for i in self.instrument_addresses.values():
+                if i.address <= 0:
+                    # No address, add empty entry
+                    measurements.append("")
+                    continue
+
+                try:
+                    # Configure Prologix to talk to the current device
+                    self.connection.write(f"++addr {i.address}")
+
+                    logging.debug(f"Triggering instrument {i.address}")
+                    # Request a single measurement
+                    val: str = self.connection.query("TRIG SGL")
+
+                    logging.debug(f"Address {i.address} Value {val}")
+
+                    # Value format is e.g. " 9.089320482E+00\r\n"
+                    # Occasionally there are also leading NULL bytes.
+                    val = val.strip(" \r\n").replace("\x00", "")
+                except Exception:
+                    # Issue reading from this instrument. Mark an error but continue
+                    # processing other instruments
+                    logging.exception(f"Exception reading from address {i.address}")
+                    val = self.ERROR_STRING
+                else:
+                    try:
+                        if i.measure_temp:
+                            val = str(volts_to_celcius(val))
+                    except AssertionError:
+                        # Issue converting value to temperature. Mark an error but
+                        # continue processing other instruments
+                        logging.exception(
+                            f"Exception converting value {val} to "
+                            f"temperature from address {i.address}"
+                        )
+                        val = self.ERROR_STRING
+
+                measurements.append(val)
+
+            assert len(measurements) == 3
+            # Can't specify in type system that the list is 3 long, so ignore the error
```

Comment:
My obvious solution to this is to create a new `dataclass` with `instrument_one` etc. fields, but it seems kind of heavy and inflexible in case we ever need to change the number of instruments available. Happy to do it if you think it's worthwhile though.

Reply:
Fixed in in 6587072540c6b78ab57df9356a78b84cfcdf8110